File: src/297-serialize-and-deserialize-binary-tree.py
# 297. Serialize and Deserialize Binary Tree

from collections import deque

# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.

        :type root: TreeNode
        :rtype: str
        """
        que = deque([])
        data_str = ''
        que.append(root)
        while len(que):
            head = que.popleft()
            if head:
                data_str += str(head.val)
                que.append(head.left)
                que.append(head.right)
            else:
                data_str += ' '
            data_str += ','
        return data_str


    def deserialize(self, data_str):
        """Decodes your encoded data to tree.

        :type data: str
        :rtype: TreeNode
        """
        data = data_str.split(',')[:-1]
        n = len(data)
        if n == 0:
            return None
        nodes = [None] * n
        for i in range(n):
            if data[i] != ' ':
                nodes[i] = TreeNode(int(data[i]))
        que = deque([])
        que.append(nodes[0])
        i = 1
        while len(que):
            head = que.popleft()
            if head:
                if i >= n:
                    break
                head.left = nodes[i]
                que.append(nodes[i])
                i += 1
                if i >= n:
                    break
                head.right = nodes[i]
                que.append(nodes[i])
                i += 1

        return nodes[0]

# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.deserialize(codec.serialize(root))

if __name__ == "__main__":
    import time
    start = time.clock()

    a = TreeNode(-1)
    b = TreeNode(0)
    c = TreeNode(1)
    d = TreeNode(4)
    e = TreeNode(5)
    f = TreeNode(6)
    g = TreeNode(7)
    a.left = b
    a.right = c
    c.left = d
    c.right = e
    d.left = f
    d.right = g

    codec = Codec()
    codes = codec.serialize(a)
    root = codec.deserialize(codes)

    end = time.clock()
    print('Running time: %s ms' % ((end - start) * 1000))

File: src/297-serialize-and-deserialize-binary-tree.cpp
// 297. Serialize and Deserialize Binary Tree


#include <iostream>
#include <vector>
#include <string>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
	int val;
	TreeNode *left;
	TreeNode *right;
	TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
 
class Codec {
public:

	// Encodes a tree to a single string.
	string serialize(TreeNode* root) {
		string result = "";
		if (root) {
			result += to_string(root->val);
			serialize_helper(root->left, result);
			serialize_helper(root->right, result);
		}
		return result;
	}

	// Decodes your encoded data to tree.
	TreeNode* deserialize(string data) {
		int start = 0, end = data.length() - 1;
		return deserialize_helper(data, start, end);
	}

private:
	void serialize_helper(TreeNode* root, string& str) {
		str += (",");
		if (root) {
			str += (to_string(root->val));
			serialize_helper(root->left, str);
			serialize_helper(root->right, str);
		}
		else {
			str += ("null");
		}
	}
	TreeNode* deserialize_helper(const string& str, int& start, int end) {
		string val_str = "";
		while (start <= end && str[start] != ',') {
			val_str += str[start];
			++start;
		}
		++start;
		TreeNode* root = nullptr;
		if (!val_str.empty() && val_str != "null") {
			root = new TreeNode(stoi(val_str));
			root->left = deserialize_helper(str, start, end);
			root->right = deserialize_helper(str, start, end);
		}
		return root;
	}
};

// Your Codec object will be instantiated and called as such:
// Codec codec;
// codec.deserialize(codec.serialize(root));

//int main() {
//	Codec codec;
//
//	TreeNode* a = new TreeNode(1);
//	TreeNode* b = new TreeNode(2);
//	TreeNode* c = new TreeNode(3);
//	TreeNode* d = new TreeNode(4);
//	a->left = b; a->right = c; c->left = d;
//
//	TreeNode* e = codec.deserialize(codec.serialize(a));
//
//	system("pause");
//	return 0;
//}